crates/sdk/src/client.rs

```diff
@@ -0,0 +1,126 @@
+use crate::{
+    local::{LocalProver, LocalProverBuilder},
+    network::{NetworkProver, NetworkProverBuilder},
+    network_v2::DEFAULT_PROVER_NETWORK_RPC,
+    opts::ProofOpts,
+    proof::SP1ProofWithPublicValues,
+    prover::Prover,
+};
+use anyhow::Result;
+use sp1_core_executor::{ExecutionError, ExecutionReport};
+use sp1_core_machine::io::SP1Stdin;
+use sp1_primitives::io::SP1PublicValues;
+use sp1_prover::{SP1ProvingKey, SP1VerifyingKey};
+use std::env;
+
+pub struct None;
+
+pub struct ProverClient {
+    inner: Box<dyn Prover>,
+}
+
+pub struct ProverClientBuilder<T> {
+    inner_builder: T,
+}
+
+impl ProverClient {
+    pub fn builder() -> ProverClientBuilder<None> {
+        ProverClientBuilder { inner_builder: None }
+    }
+
+    #[deprecated(note = "Use ProverClient::builder() instead")]
+    pub fn new() -> Self {
+        Self::create_from_env()
+    }
+
+    fn create_from_env() -> Self {
+        match env::var("SP1_PROVER").unwrap_or("local".to_string()).as_str() {
+            "network" => {
+                let rpc_url = env::var("PROVER_NETWORK_RPC")
+                    .unwrap_or_else(|_| DEFAULT_PROVER_NETWORK_RPC.to_string());
+                let private_key = env::var("SP1_PRIVATE_KEY").unwrap_or_default();
+
+                let network_prover = NetworkProver::new(rpc_url, private_key);
+                ProverClient { inner: Box::new(network_prover) }
+            }
+            _ => {
+                let local_prover = LocalProver::new();
+                ProverClient { inner: Box::new(local_prover) }
+            }
+        }
+    }
+
+    pub async fn setup(&self, elf: &[u8]) -> (SP1ProvingKey, SP1VerifyingKey) {
+        self.inner.setup(elf).await
+    }
+
+    pub async fn execute(
+        &self,
+        elf: &[u8],
+        stdin: &SP1Stdin,
+    ) -> Result<(SP1PublicValues, ExecutionReport), ExecutionError> {
+        self.inner.execute(elf, stdin).await
+    }
+
+    pub async fn prove(
```

Comment:
This should return a `DynProofRequest` which lets you set options like `prove(..).groth16()` as discussed yesterday

Reply:
updated